maintenance/annotation_maint.py

```diff
@@ -1,38 +1,147 @@
 import allel
-import zarr
+import operator
+import pathlib
 import numpy as np
-import stdpopsim as stp
 import logging
-import warnings
-import urllib.request
 import os
+import glob
+import tarfile
+import stdpopsim
 
 logger = logging.getLogger(__name__)
 # make root directory for zarr annotations
 annot_path = "annotations"
 os.mkdir(annot_path)
+
+
+def merged(intervals, *, closed: bool):
+    """
+    Merge overlapping and adjacent intervals.
+
+    :param intervals: An iterable of (start, end) coordinates.
+    :param bool closed: If True, [start, end] coordinates are closed,
+        so [1, 2] and [3, 4] are adjacent intervals and will be merged.
+        If False, [start, end) coordinates are half-open,
+        so [1, 2) and [3, 4) are not adjacent and will not be merged.
+    """
+
+    def iter_merged(intervals, *, closed: bool):
+        """
+        Generate tuples of (start, end) coordinates for merged intervals.
+        """
+        intervals = sorted(intervals, key=operator.itemgetter(0))
+        if len(intervals) == 0:
+            return
+        start, end = intervals[0]
+        for a, b in intervals[1:]:
+            assert a <= b
+            if a > end + closed:
+                # No intersection with the current interval.
+                yield start, end
+                start, end = a, b
+            else:
+                # Intersects, or is contiguous with, the current interval.
+                end = max(end, b)
+        yield start, end
+
+    return list(iter_merged(intervals, closed=closed))
+
+
+def test_merged():
```

Comment:
The test should go in `tests/test_maintenance.py`.

Reply:
i think `tests/test_annotation.py` is actually the right place

Reply:
Ok, but the test function can be removed from here now.